MatMul heuristics for aarch64

yd2102 opened the pull request:
This PR focuses on improving MatMul performance for aarch64 only. It introduces a light-weight heuristic that dispatches small or tall/flat MatMul operations to OpenBLAS while other shapes to MKLDNN/ACL.

On average, the proposed heuristics improve MatMul operator latency by 1.03x / 1.04x / 1.05x / 1.09x / 1.22x for 1 / 2 / 4 / 8 / 16 threads, respectively (baseline is using ACL for all MatMuls on AWS Graviton c7g instances).

Fixes #107168 

<details>
<summary>Full MatMul benchmark script and result</summary>

Run this following script `run.sh` with `ABt.py` under the same directory:
```shell
#!/bin/bash

script=ABt.py
prefix=acl

OMP_NUM_THREADS=1 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=0 TORCH_MKLDNN_MATMUL_MIN_SIZE=0 python ${script} > ${prefix}_t1.csv
OMP_NUM_THREADS=2 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=0 TORCH_MKLDNN_MATMUL_MIN_SIZE=0 python ${script} > ${prefix}_t2.csv
OMP_NUM_THREADS=4 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=0 TORCH_MKLDNN_MATMUL_MIN_SIZE=0 python ${script} > ${prefix}_t4.csv
OMP_NUM_THREADS=8 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=0 TORCH_MKLDNN_MATMUL_MIN_SIZE=0 python ${script} > ${prefix}_t8.csv
OMP_NUM_THREADS=16 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=0 TORCH_MKLDNN_MATMUL_MIN_SIZE=0 python ${script} > ${prefix}_t16.csv

prefix=heur
OMP_NUM_THREADS=1 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=8 TORCH_MKLDNN_MATMUL_MIN_SIZE=8192 python ${script} > ${prefix}_t1.csv
OMP_NUM_THREADS=2 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=8 TORCH_MKLDNN_MATMUL_MIN_SIZE=8192 python ${script} > ${prefix}_t2.csv
OMP_NUM_THREADS=4 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=8 TORCH_MKLDNN_MATMUL_MIN_SIZE=8192 python ${script} > ${prefix}_t4.csv
OMP_NUM_THREADS=8 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=8 TORCH_MKLDNN_MATMUL_MIN_SIZE=8192 python ${script} > ${prefix}_t8.csv
OMP_NUM_THREADS=16 DNNL_DEFAULT_FPMATH_MODE=BF16 TORCH_MKLDNN_MATMUL_MIN_DIM=8 TORCH_MKLDNN_MATMUL_MIN_SIZE=8192 python ${script} > ${prefix}_t16.csv
```

`ABt.py`:
```python
import argparse
import timeit
import torch
import numpy as np


BATCH = 1
DIM_MIN = 8
DIM_MAX = 256

M_MIN = DIM_MIN
K_MIN = DIM_MIN
N_MIN = DIM_MIN

M_MAX = DIM_MAX
K_MAX = DIM_MAX
N_MAX = DIM_MAX

min_iterations = 1000
min_time = 0.1

def get_stats(timings):
    times = np.array(timings)
    time_avg = np.average(times) * 1000
    time_med = np.median(times) * 1000
    time_90th = np.percentile(times, 90) * 1000
    time_99th = np.percentile(times, 99) * 1000

    return time_avg, time_med, time_90th, time_99th


def bench(M, K, N, min_iterations, min_time):
    a = torch.randn(M, K, dtype=torch.float32)
    b = torch.randn(N, K, dtype=torch.float32)
    timings = []
    
    with torch.no_grad():
        for _ in range(max(1, min_iterations // 100)):
            c = torch.matmul(a, b.transpose(0, 1))
        
        bench_time = timeit.default_timer()
        
        while True:
            for _ in range(min_iterations):
                start_time = timeit.default_timer()
                c = torch.matmul(a, b.transpose(0, 1))
                end_time = timeit.default_timer()
                timings.append(end_time - start_time)
                
            if timeit.default_timer() - bench_time > min_time:
                break
            
    return get_stats(timings)
    
    
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('-l', '--loop', dest='loop', action='store_true')
    flags = parser.parse_args()
    
    if flags.loop:
        while True:
            for M in range(M_MAX//2, M_MAX+1, 8):
                for K in range(K_MAX//2, K_MAX+1, 8):
                    for N in range(N_MAX//2, N_MAX+1, 8):
                        stats = bench(M, K, N, min_iterations, min_time)
    else:
        torch.manual_seed(0)
        print(f"M, K, N, latency")
        
        for M in range(M_MIN, M_MAX+1, 8):
            for K in range(K_MIN, K_MAX+1, 8):
                for N in range(N_MIN, N_MAX+1, 8):
                    stats = bench(M, K, N, min_iterations, min_time)
                    print(f"{M}, {K}, {N}, {stats[2]}")
```

Here's the benchmark result collected on AWS Graviton c7g instance. Due to the size of the table, I can only attach the result in this text file:

[table.txt](https://github.com/pytorch/pytorch/files/12374265/table.txt)

</details>

cc @malfet

yd2102 committed:
apply d8_8k heuristics by default

yd2102 committed:
Merge branch 'pytorch:main' into pr_heur_batch

yd2102 committed:
add comments for default dimension and size

yd2102 committed:
add comments for default dimension and size

yd2102 committed:
Merge branch 'pr_heur_batch' of https://github.com/yd2102/pytorch into pr_heur_batch

yd2102 committed:
Merge branch 'pytorch:main' into pr_heur_batch

yd2102 committed:
stylistic change for apply_mkldnn_matmul_heur()

yd2102 commented:
Hi @nikitaved @IvanYashchuk, Could one of you kindly review this PR? Thanks!

malfet reviewed (commented):
@yd2102 please add a comment outlining what benchmark you've run on what hardware to demonstrate perf gains

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1322,6 +1322,46 @@ Tensor outer(const Tensor& self, const Tensor& vec2) {
+static inline int64_t get_mkldnn_matmul_min_dim() {
+  static auto value = [&] {
+    // Minimum dimension requirement for MKLDNN; derived based on experiments.
+    constexpr int64_t default_min_dim = 8;
```
Hmm, why not allow users to specify min_dim via environment variable on x86 as well?

yd2102 replied:
@malfet I think there are two reasons we want to keep such environment variables specific to aarch64:

1. We're trying to improve performance on aarch64, and all the analysis and benchmarking were done on aarch64 (AWS Graviton c7g). So it makes more sense not to expose such variables for x86 because that wouldn't be meaningful.
2. The other reason is it appears to me MKL on x86 doesn't go through this [mkldnn_matmul](https://github.com/pytorch/pytorch/blob/29813c61ea61d616230d3251ade81f56472ecb77/aten/src/ATen/native/mkldnn/Matmul.cpp#L121) code path (i.e. doesn't go through ideep->oneDNN). At least from the hugging-face "bert-base-uncased" model it's what I have observed. Therefore, such heuristics probably won't make sense for x86.

malfet replied:
@yd2102 just to clarify: I'm not proposing to change the default to 8 and 8k on x86, I just want to avoid code bifurcation by having the same options available across all platforms, that also keeps divergent codepath smaller (i.e. you would need 2 line divergence to set default `min_dim` rather than 20+ lines divergence you are having right now)

yd2102 replied:
@malfet I have pushed the change to converge the code paths based on your suggestion.

yd2102 commented:
> @yd2102 please add a comment outlining what benchmark you've run on what hardware to demonstrate perf gains

@malfet The benchmarking is pretty straightforward. I simply measured the inference latency of each selected Hugging-Face model using inputs of various lengths. Something like this:
```
import timeit
timings = []
for _ in range(iterations):
    start_time = timeit.default_timer()
    output = model("some_input_text")
    end_time = timeit.default_timer()
    timings.append(end_time - start_time)
```
For completeness, I appended to [this comment](https://github.com/pytorch/pytorch/issues/107168#issue-1850427960) a full list of inputs for each model.

The hardware that I ran these benchmarks on was AWS Graviton c7g instance.

malfet commented:
> @malfet The benchmarking is pretty straightforward. I simply measured the inference latency of each selected Hugging-Face model using inputs of various lengths. Something like this:
> 
> ```
> import timeit
> timings = []
> for _ in range(iterations):
>     start_time = timeit.default_timer()
>     output = model("some_input_text")
>     end_time = timeit.default_timer()
>     timings.append(end_time - start_time)
> ```

@yd2102 Would it be possible to  reduce the benchmark to just matmul calls (as hugging face models can change with time, but it would be nice to keep benchmarks reproducible say two years from now)

In either case, please specify complete benchmark script as part of PR description as well as perf number on the platform of your choice before and after the change

> 
> For completeness, I appended to [this comment](https://github.com/pytorch/pytorch/issues/107168#issue-1850427960) a full list of inputs for each model.
> 
> The hardware that I ran these benchmarks on was AWS Graviton c7g instance.

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1322,6 +1322,46 @@ Tensor outer(const Tensor& self, const Tensor& vec2) {
+  return value;
+}
+
+static inline bool apply_mkldnn_matmul_heur(int64_t m, int64_t k, int64_t n, int64_t min_dim, int64_t min_size) {
+  return m > min_dim && k > min_dim && n > min_dim && m * k * n > min_size;
```
Q: why explicitly pass `min_dim`/`min_size` arguments when they can only be by calling the respective functions, which also have no other use. 
```suggestion
static inline bool apply_mkldnn_matmul_heur(int64_t m, int64_t k, int64_t n) {
  const auto min_dim = get_mkldnn_matmul_min_dim();
  const auto min_size = get_mkldnn_matmul_min_size();
  return m > min_dim && k > min_dim && n > min_dim && m * k * n > min_size;
```

yd2102 replied:
@malfet I changed the function signature based on your feedback.

malfet reviewed (changes requested):
Please address the feedback:
- specify full reproducible benchmark and observed perf gains in the PR description
- Keep platform specific code divergence at minimum, by keeping implementation of functions the same, but using different defaults for x86 and arm platforms

yd2102 committed:
Consolidate x86 and aarch64 code paths

yd2102 commented:
> Please address the feedback:
> 
> * specify full reproducible benchmark and observed perf gains in the PR description
> * Keep platform specific code divergence at minimum, by keeping implementation of functions the same, but using different defaults for x86 and arm platforms

@malfet  I have addressed your feedback as follows:
1. The perf gain, full MatMul benchmark result, and the scripts are put in this PR description [above](https://github.com/pytorch/pytorch/pull/107167#issue-1850426364).
2. Done.

yd2102 commented:
> > For completeness, I appended to [this comment](https://github.com/pytorch/pytorch/issues/107168#issue-1850427960) a full list of inputs for each model.
> 
> Perhaps I'm reading it wrong, but this comment lacks the baseline performance, isn't it?

The baseline is using MKLDNN/ACL for all MatMul operations. I updated this PR description [above](https://github.com/pytorch/pytorch/pull/107167#issue-1850426364) to make it more clear. Thanks.

yd2102 committed:
Use positive heuristic default value only if arch is Neoverse V1 (AWS c7g)

malfet commented:
@yd2102 thank you for the updates

malfet commented:
@pytorchbot merge
